pkg/tfgen/docs.go

```diff
@@ -867,69 +876,259 @@ func getNestedBlockName(line string) string {
+
 func parseArgReferenceSection(subsection []string, ret *entityDocs) {
-	var lastMatch string
-	var nested docsPath
-
-	addNewHeading := func(name, desc, line string) {
-		// found a property bullet, extract the name and description
-		if nested != "" {
-			// We found this line within a nested field. We should record it as such.
-			if ret.Arguments[nested] == nil {
-				totalArgumentsFromDocs++
+	ret.ensure()
+	src := []byte(strings.Join(subsection, "\n"))
+	node := goldmark.New().Parser().Parse(
+		text.NewReader(src))
+
+	type namedNode struct {
+		node     ast.Node
+		name     string
+		writable bool
+	}
+	var parent, child, latest, latestParent *namedNode
+	setLatest := func() { latest, latestParent = child, parent }
+
+	skipPrint := map[ast.Node]bool{}
+
+	baked := map[docsPath]struct{}{}
+	unbaked := map[docsPath]struct{}{}
+
+	// The algorithm for this parse works on a single descend + ascend pass.
+	//
+	// During descent, significant nodes such as new headings are marked. During
+	// ascent, nodes are written to ret.
+
+	descend := func(node ast.Node) ast.WalkStatus {
+		switch node.Kind() {
+		case ast.KindThematicBreak:
+			for k := range unbaked {
+				baked[k] = struct{}{}
 			}
-			ret.Arguments[nested.join(name)] = &argumentDocs{desc}
-		} else {
-			if genericNestedRegexp.MatchString(line) {
-				return
+			unbaked = map[docsPath]struct{}{}
+		case ast.KindListItem:
+			// We have entered a new item
+			if name := nodeItemName(src, node); name != "" && child == nil {
+				var key docsPath
+				if parent == nil {
+					child = nil
+					parent = &namedNode{node, name, true}
+					key = docsPath(name)
+				} else {
+					child = &namedNode{node, name, true}
+					key = docsPath(parent.name).join(name)
+				}
+				setLatest()
+
+				_, isBaked := baked[key]
+				if v, ok := ret.Arguments[key]; ok && !isBaked {
+					v.description = ""
+				}
+				unbaked[key] = struct{}{}
+			}
+		case ast.KindHeading:
+			target := isHeadingTarget(src, node.(*ast.Heading))
+			if target != "" {
+				child = &namedNode{node, target, true}
+				parent = child
+				setLatest()
+				return ast.WalkSkipChildren
+			}
+		case ast.KindCodeSpan, ast.KindCodeBlock:
+			if nxt := node.NextSibling(); nxt != nil && nxt.Kind() == ast.KindText {
+				// Observe annotations like:
+				//
+				//	Inside of `dns` section the following argument are supported:
+				//
+				// When we find these, it overrides the current node.
+				txt := fmt.Sprintf("`%s` %s", string(node.Text(src)),
+					string(nxt.Text(src)))
+				target := getNestedBlockName([]byte(txt))
+				if target != "" {
+					// Setting both child an parent effectively sets
+					// this until the list ends (and we reset).
+					p := node.Parent()
+					child = &namedNode{p, target, false}
+					parent = child
+					setLatest()
+					return ast.WalkSkipChildren
+				}
+			}
+		case ast.KindText:
+			if genericNestedRegexp.Match(node.Text(src)) {
```

Comment:
What does node.Text() do?  It's been a while since I touched goldmark, I more recently wrestled with blackfriday. It can be really difficult to work with.

Reply:
`node`s are flyweight constructs. They don't have access to the text that they represent. `node.Text(src []byte) []byte` returns the text from `src` that `node` represents.